### ISSUE TITLE
how to set "eagle_choices" or "medusa_choices" per request

### ISSUE THREAD

**Issue by avianion** [NVIDIA/TensorRT-LLM#2522]
I have seen in several comments that it is possible or should  be possible to set eagle_choices or medusa_choices per request.

Where can this be done when using the executor API? is there a specific file or method to use to do this? Because when using the executor, it seems you set them once and they cannot be changed for that runtime session.

please point me in the right direction.

@kaiyux @AdamzNV @byshiue @nekorobov

**Comment by nekorobov** [comment]
Hi @avianion, medusa_choices can be set only per server, i.e. the same tree for all requests. This is the limitation of current implementation. 
For Eagle, you can set them per request using [EagleConfig](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/include/tensorrt_llm/executor/executor.h#L614C10-L614C24).

**Comment by gabriel-peracio** [comment]
Is it possible to entirely disable medusa/eagle per request?

**Comment by nekorobov** [comment]
No, either all requests have to be eagle/medusa or none of them.

**Comment by avianion** [comment]
@nekorobov what I mean is, lets say you have a batch of 16 requests, i want to have a draft length of 63 for those, then for another batch of 8, lets say I want to have a draft length of 1. it should be possible, right?

**Comment by nekorobov** [comment]
Yes, this is possible in Eagle. In Medusa the tree is set in the initialization and is fixed during the lifetime of the server. Another limitation of the Eagle is that the tree of all requests must have the same depth equal to the `--num_eagle_layers` specified during the `convert_checkpoint.py`

**Comment by avianion** [comment]
@nekorobov 

```
            auto executorRequests = createExecutorRequests(request, mInstanceSpecificConfig.excludeInputFromOutput,
                isDecoupled(), mModelType, mIsOrchestratorMode, mSpeculativeDecodingFastLogits);
            if (true) {
                std::vector<std::vector<int>> choices = {{0}, {0,0}, {0,0,0}, {0,0,0,0}};
                tensorrt_llm::executor::EagleConfig eagleConfig(choices);
                // Set EagleConfig for each executor request in the batch
                for (auto& execRequest : executorRequests) {
                    execRequest.setEagleConfig(eagleConfig);
                }
            }

```

here is how im manually setting the eagleConfig on line 804 of model_instance_state.cc (triton server)

does this look correct to you or no?

**Comment by karljang** [comment]
Closing the issue with the hope that the discussions above addressed your question.
Please feel free to open a new one~
Thank you!
